Userland/Libraries/LibJS/AST.cpp

```diff
@@ -444,129 +475,274 @@ Value ForStatement::execute(Interpreter& interpreter, GlobalObject& global_objec
+                return throw_completion(exception->value());
+            return {};
         }
-        variable_declaration.execute(interpreter, global_object);
-        return variable_declaration.declarations().first().target();
+        if (lhs_kind == Assignment) {
+            VERIFY(pattern_lhs);
+            return interpreter.vm().destructuring_assignment_evaluation(*pattern_lhs, next_value, global_object);
+        }
+        VERIFY(expression_lhs && is<VariableDeclaration>(*expression_lhs));
+        auto& for_declaration = static_cast<VariableDeclaration const&>(*expression_lhs);
+        auto& binding_pattern = for_declaration.declarations().first().target().get<NonnullRefPtr<BindingPattern>>();
+        VERIFY(lhs_kind == VarBinding || iteration_environment);
+        return interpreter.vm().binding_initialization(binding_pattern, next_value, iteration_environment, global_object);
     }
+};
+
+// 14.7.5.6 ForIn/OfHeadEvaluation ( uninitializedBoundNames, expr, iterationKind ), https://tc39.es/ecma262/#sec-runtime-semantics-forinofheadevaluation
+static ThrowCompletionOr<ForInOfHeadState> for_in_of_head_execute(Interpreter& interpreter, GlobalObject& global_object, Variant<NonnullRefPtr<ASTNode>, NonnullRefPtr<BindingPattern>> lhs, Expression const& rhs)
+{
+
+    ForInOfHeadState state(lhs);
+    if (auto* ast_ptr = lhs.get_pointer<NonnullRefPtr<ASTNode>>(); ast_ptr && is<VariableDeclaration>(*(*ast_ptr))) {
+        Environment* new_environment = nullptr;
+
+        auto& variable_declaration = static_cast<VariableDeclaration const&>(*(*ast_ptr));
+        VERIFY(variable_declaration.declarations().size() == 1);
+        state.destructuring = variable_declaration.declarations().first().target().has<NonnullRefPtr<BindingPattern>>();
+        if (variable_declaration.declaration_kind() == DeclarationKind::Var) {
+            state.lhs_kind = ForInOfHeadState::VarBinding;
+            auto& variable = variable_declaration.declarations().first();
+            // B.3.5 Initializers in ForIn Statement Heads, https://tc39.es/ecma262/#sec-initializers-in-forin-statement-heads
+            if (variable.init()) {
+                VERIFY(variable.target().has<NonnullRefPtr<Identifier>>());
+                auto& binding_id = variable.target().get<NonnullRefPtr<Identifier>>()->string();
+                auto reference = interpreter.vm().resolve_binding(binding_id);
+                if (auto* exception = interpreter.exception())
+                    return throw_completion(exception->value());
+
+                auto result = TRY(interpreter.vm().named_evaluation_if_anonymous_function(global_object, *variable.init(), binding_id));
+                reference.put_value(global_object, result);
+                if (auto* exception = interpreter.exception())
+                    return throw_completion(exception->value());
+            }
+        } else {
+            state.lhs_kind = ForInOfHeadState::LexicalBinding;
```

Comment:
This seems to match "Step 2: If uninitializedBoundNames is not an empty List, then" in the 14.7.5.6 spec.
But I don't understand _at all_ why you're doing that after checking whether `variable_declaration.declaration_kind() == DeclarationKind::Var` instead of checking if `variable_declaration.declarations().is_empty()`, like the spec seems to require?

Edit: In general, I don't really understand what this is supposed to match in the spec... THis part kinda matches step 2, but the rest of the method doesn't seem to match anything in 14.7.5.6 (unless I really don't understand the code)...

Reply:
Yeah a similar comments as with the other this was a weird combination of spec which is hopefully a little more clear with the   comments now.